src/Mono.Android/Android.Runtime/JavaProxyThrowable.cs

```diff
@@ -1,23 +1,83 @@
 using System;
+using System.Diagnostics;
+using System.Reflection;
+
+using StackTraceElement = Java.Lang.StackTraceElement;
 
 namespace Android.Runtime {
 
-	class JavaProxyThrowable : Java.Lang.Error {
+	sealed class JavaProxyThrowable : Java.Lang.Error {
 
 		public  readonly Exception InnerException;
 
-		public JavaProxyThrowable (Exception innerException)
-			: base (GetDetailMessage (innerException))
+		JavaProxyThrowable (string message, Exception innerException)
+			: base (message)
 		{
 			InnerException  = innerException;
 		}
 
-		static string GetDetailMessage (Exception innerException)
+		public static JavaProxyThrowable Create (Exception innerException)
+		{
+			if (innerException == null) {
+				throw new ArgumentNullException (nameof (innerException));
+			}
+
+			// We prepend managed exception type to message since Java will see `JavaProxyThrowable` instead.
+			var proxy = new JavaProxyThrowable ($"[{innerException.GetType}]: {innerException.Message}", innerException);
+
+			try {
+				proxy.TranslateStackTrace ();
+			} catch (Exception ex) {
+				// We shouldn't throw here, just try to do the best we can do
+				Console.WriteLine ($"JavaProxyThrowable: translation threw an exception: {ex}");
+				proxy = new JavaProxyThrowable (innerException.ToString (), innerException);
+			}
+
+			return proxy;
+		}
+
+		void TranslateStackTrace ()
 		{
-			if (innerException == null)
-				throw new ArgumentNullException ("innerException");
+			var trace = new StackTrace (InnerException, fNeedFileInfo: true);
+			if (trace.FrameCount <= 0) {
+				return;
+			}
+
+			StackTraceElement[]? javaTrace = null;
+			try {
+				javaTrace = GetStackTrace ();
+			} catch (Exception ex) {
+				// Report...
+				Console.WriteLine ($"JavaProxyThrowable: obtaining Java stack trace threw an exception: {ex}");
+				// ..but ignore
+			}
+
+
+			StackFrame[] frames = trace.GetFrames ();
+			int nElements = frames.Length + (javaTrace?.Length ?? 0);
+			StackTraceElement[] elements = new StackTraceElement[nElements];
+
+			for (int i = 0; i < frames.Length; i++) {
+				StackFrame managedFrame = frames[i];
+				MethodBase? managedMethod = managedFrame.GetMethod ();
+
+				var throwableFrame = new StackTraceElement (
+					declaringClass: managedMethod?.DeclaringType?.FullName,
+					methodName: managedMethod?.Name,
+					fileName: managedFrame?.GetFileName (),
+					lineNumber: managedFrame?.GetFileLineNumber () ?? 0
```

Comment:
This should probably be `?? -1`, as negative numbers appear to be the "don't have one" behavior.  See e.g.

```
W *jonp*  : # jonp: stack[3]=crc64ef325ad1a3b68b57.MainActivity.n_onCreate(MainActivity.java:-2)
```

No idea where `-2` came from; I assume it's because the file has no line number info?  (Which also makes no sense, as the declaration for `n_onCreate()` *does* have a line number…)